// @aztec/aztec.js@0.7.10/dest/account/defaults/default_entrypoint.js
import { Fr, FunctionData, TxContext } from '@aztec/circuits.js';
import { encodeArguments } from '@aztec/foundation/abi';
import { PackedArguments, TxExecutionRequest } from '@aztec/types';
import { DEFAULT_CHAIN_ID, DEFAULT_VERSION } from '../../utils/defaults.js';
import { buildPayload, hashPayload } from './entrypoint_payload.js';
/**
 * Implementation for an entrypoint interface that follows the default entrypoint signature
 * for an account, which accepts an EntrypointPayload as defined in noir-libs/aztec-noir/src/entrypoint.nr.
 */
export class DefaultAccountEntrypoint {
    constructor(address, auth, chainId = DEFAULT_CHAIN_ID, version = DEFAULT_VERSION) {
        this.address = address;
        this.auth = auth;
        this.chainId = chainId;
        this.version = version;
    }
    async createTxExecutionRequest(executions) {
        const { payload, packedArguments: callsPackedArguments } = await buildPayload(executions);
        const abi = this.getEntrypointAbi();
        const packedArgs = await PackedArguments.fromArgs(encodeArguments(abi, [payload]));
        const message = Fr.fromBuffer(await hashPayload(payload));
        const authWitness = await this.auth.createAuthWitness(message);
        const txRequest = TxExecutionRequest.from({
            argsHash: packedArgs.hash,
            origin: this.address,
            functionData: FunctionData.fromAbi(abi),
            txContext: TxContext.empty(this.chainId, this.version),
            packedArguments: [...callsPackedArguments, packedArgs],
            authWitnesses: [authWitness],
        });
        return txRequest;
    }
    getEntrypointAbi() {
        return {
            name: 'entrypoint',
            functionType: 'secret',
            isInternal: false,
            parameters: [
                {
                    name: 'payload',
                    type: {
                        kind: 'struct',
                        path: 'aztec::entrypoint::EntrypointPayload',
                        fields: [
                            {
                                name: 'function_calls',
                                type: {
                                    kind: 'array',
                                    length: 4,
                                    type: {
                                        kind: 'struct',
                                        path: 'aztec::entrypoint::FunctionCall',
                                        fields: [
                                            {
                                                name: 'args_hash',
                                                type: {
                                                    kind: 'field',
                                                },
                                            },
                                            {
                                                name: 'function_selector',
                                                type: {
                                                    kind: 'field',
                                                },
                                            },
                                            {
                                                name: 'target_address',
                                                type: {
                                                    kind: 'field',
                                                },
                                            },
                                            {
                                                name: 'is_public',
                                                type: {
                                                    kind: 'boolean',
                                                },
                                            },
                                        ],
                                    },
                                },
                            },
                            {
                                name: 'nonce',
                                type: {
                                    kind: 'field',
                                },
                            },
                        ],
                    },
                    visibility: 'public',
                },
            ],
            returnTypes: [],
        };
    }
}
