// @aztec/aztec.js@0.7.10/dest/account/contract/ecdsa_account_contract.js
import { Ecdsa } from '@aztec/circuits.js/barretenberg';
import { AuthWitness } from '@aztec/types';
import EcdsaAccountContractAbi from '../../abis/ecdsa_account_contract.json' assert { type: 'json' };
import { BaseAccountContract } from './base_account_contract.js';
/**
 * Account contract that authenticates transactions using ECDSA signatures
 * verified against a secp256k1 public key stored in an immutable encrypted note.
 */
export class EcdsaAccountContract extends BaseAccountContract {
    constructor(signingPrivateKey) {
        super(EcdsaAccountContractAbi);
        this.signingPrivateKey = signingPrivateKey;
    }
    async getDeploymentArgs() {
        const signingPublicKey = await Ecdsa.new().then(e => e.computePublicKey(this.signingPrivateKey));
        return [signingPublicKey.subarray(0, 32), signingPublicKey.subarray(32, 64)];
    }
    getAuthWitnessProvider(_address) {
        return new EcdsaAuthWitnessProvider(this.signingPrivateKey);
    }
}
/** Creates auth witnesses using ECDSA signatures. */
class EcdsaAuthWitnessProvider {
    constructor(signingPrivateKey) {
        this.signingPrivateKey = signingPrivateKey;
    }
    async createAuthWitness(message) {
        const ecdsa = await Ecdsa.new();
        const signature = ecdsa.constructSignature(message.toBuffer(), this.signingPrivateKey);
        return new AuthWitness(message, [...signature.r, ...signature.s]);
    }
}
