// @aztec/aztec.js@0.7.10/dest/account/contract/schnorr_account_contract.js
import { Schnorr } from '@aztec/circuits.js/barretenberg';
import { AuthWitness } from '@aztec/types';
import SchnorrAccountContractAbi from '../../abis/schnorr_account_contract.json' assert { type: 'json' };
import { BaseAccountContract } from './base_account_contract.js';
/**
 * Account contract that authenticates transactions using Schnorr signatures
 * verified against a Grumpkin public key stored in an immutable encrypted note.
 */
export class SchnorrAccountContract extends BaseAccountContract {
    constructor(signingPrivateKey) {
        super(SchnorrAccountContractAbi);
        this.signingPrivateKey = signingPrivateKey;
    }
    async getDeploymentArgs() {
        const signingPublicKey = await Schnorr.new().then(e => e.computePublicKey(this.signingPrivateKey));
        return [signingPublicKey.x, signingPublicKey.y];
    }
    getAuthWitnessProvider(_address) {
        return new SchnorrAuthWitnessProvider(this.signingPrivateKey);
    }
}
/** Creates auth witnesses using Schnorr signatures. */
class SchnorrAuthWitnessProvider {
    constructor(signingPrivateKey) {
        this.signingPrivateKey = signingPrivateKey;
    }
    async createAuthWitness(message) {
        const schnorr = await Schnorr.new();
        const signature = schnorr.constructSignature(message.toBuffer(), this.signingPrivateKey).toBuffer();
        return new AuthWitness(message, [...signature]);
    }
}
