// @aztec/circuits.js@0.7.10/dest/structs/rollup/base_rollup.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Class containing the data of a preimage of a single leaf in the nullifier tree.
 * Note: It's called preimage because this data gets hashed before being inserted as a node into the `IndexedTree`.
 */
export class NullifierLeafPreimage {
    constructor(
    /**
     * Leaf value inside the indexed tree's linked list.
     */
    leafValue, 
    /**
     * Next value inside the indexed tree's linked list.
     */
    nextValue, 
    /**
     * Index of the next leaf in the indexed tree's linked list.
     */
    nextIndex) {
        this.leafValue = leafValue;
        this.nextValue = nextValue;
        this.nextIndex = nextIndex;
    }
    toBuffer() {
        return serializeToBuffer(this.leafValue, this.nextValue, this.nextIndex);
    }
    static empty() {
        return new NullifierLeafPreimage(Fr.ZERO, Fr.ZERO, 0);
    }
}
/**
 * Data which is forwarded through the base rollup circuits unchanged.
 */
export class ConstantRollupData {
    constructor(
    /**
     * Snapshot of the historic blocks roots tree at the start of the rollup.
     */
    startHistoricBlocksTreeRootsSnapshot, 
    /**
     * Root of the private kernel verification key tree.
     */
    privateKernelVkTreeRoot, 
    /**
     * Root of the public kernel circuit verification key tree.
     */
    publicKernelVkTreeRoot, 
    /**
     * Hash of the base rollup circuit verification key.
     */
    baseRollupVkHash, 
    /**
     * Hash of the merge rollup circuit verification key.
     */
    mergeRollupVkHash, 
    /**
     * Global variables for the block
     */
    globalVariables) {
        this.startHistoricBlocksTreeRootsSnapshot = startHistoricBlocksTreeRootsSnapshot;
        this.privateKernelVkTreeRoot = privateKernelVkTreeRoot;
        this.publicKernelVkTreeRoot = publicKernelVkTreeRoot;
        this.baseRollupVkHash = baseRollupVkHash;
        this.mergeRollupVkHash = mergeRollupVkHash;
        this.globalVariables = globalVariables;
    }
    static from(fields) {
        return new ConstantRollupData(...ConstantRollupData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ConstantRollupData(reader.readObject(AppendOnlyTreeSnapshot), reader.readFr(), reader.readFr(), reader.readFr(), reader.readFr(), reader.readObject(GlobalVariables));
    }
    static getFields(fields) {
        return [
            fields.startHistoricBlocksTreeRootsSnapshot,
            fields.privateKernelVkTreeRoot,
            fields.publicKernelVkTreeRoot,
            fields.baseRollupVkHash,
            fields.mergeRollupVkHash,
            fields.globalVariables,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...ConstantRollupData.getFields(this));
    }
}
/**
 * Inputs to the base rollup circuit.
 */
export class BaseRollupInputs {
    constructor(
    /**
     * Data of the 2 kernels that preceded this base rollup circuit.
     */
    kernelData, 
    /**
     * Snapshot of the private data tree at the start of the base rollup circuit.
     */
    startPrivateDataTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the start of the base rollup circuit.
     */
    startNullifierTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the start of the base rollup circuit.
     */
    startContractTreeSnapshot, 
    /**
     * Root of the public data tree at the start of the base rollup circuit.
     */
    startPublicDataTreeRoot, 
    /**
     * Snapshot of the historic blocks tree at the start of the base rollup circuit.
     */
    startHistoricBlocksTreeSnapshot, 
    /**
     * The nullifiers which need to be updated to perform the batch insertion of the new nullifiers.
     * See `StandardIndexedTree.batchInsert` function for more details.
     */
    lowNullifierLeafPreimages, 
    /**
     * Membership witnesses for the nullifiers which need to be updated to perform the batch insertion of the new
     * nullifiers.
     */
    lowNullifierMembershipWitness, 
    /**
     * Sibling path "pointing to" where the new commitments subtree should be inserted into the private data tree.
     */
    newCommitmentsSubtreeSiblingPath, 
    /**
     * Sibling path "pointing to" where the new nullifiers subtree should be inserted into the nullifier tree.
     */
    newNullifiersSubtreeSiblingPath, 
    /**
     * Sibling path "pointing to" where the new contracts subtree should be inserted into the contract tree.
     */
    newContractsSubtreeSiblingPath, 
    /**
     * Sibling paths of leaves which are to be affected by the public data update requests.
     * Each item in the array is the sibling path that corresponds to an update request.
     */
    newPublicDataUpdateRequestsSiblingPaths, 
    /**
     * Sibling paths of leaves which are to be read by the public data reads.
     * Each item in the array is the sibling path that corresponds to a read request.
     */
    newPublicDataReadsSiblingPaths, 
    /**
     * Membership witnesses of historic blocks referred by each of the 2 kernels.
     */
    historicBlocksTreeRootMembershipWitnesses, 
    /**
     * Data which is not modified by the base rollup circuit.
     */
    constants) {
        this.kernelData = kernelData;
        this.startPrivateDataTreeSnapshot = startPrivateDataTreeSnapshot;
        this.startNullifierTreeSnapshot = startNullifierTreeSnapshot;
        this.startContractTreeSnapshot = startContractTreeSnapshot;
        this.startPublicDataTreeRoot = startPublicDataTreeRoot;
        this.startHistoricBlocksTreeSnapshot = startHistoricBlocksTreeSnapshot;
        this.lowNullifierLeafPreimages = lowNullifierLeafPreimages;
        this.lowNullifierMembershipWitness = lowNullifierMembershipWitness;
        this.newCommitmentsSubtreeSiblingPath = newCommitmentsSubtreeSiblingPath;
        this.newNullifiersSubtreeSiblingPath = newNullifiersSubtreeSiblingPath;
        this.newContractsSubtreeSiblingPath = newContractsSubtreeSiblingPath;
        this.newPublicDataUpdateRequestsSiblingPaths = newPublicDataUpdateRequestsSiblingPaths;
        this.newPublicDataReadsSiblingPaths = newPublicDataReadsSiblingPaths;
        this.historicBlocksTreeRootMembershipWitnesses = historicBlocksTreeRootMembershipWitnesses;
        this.constants = constants;
    }
    static from(fields) {
        return new BaseRollupInputs(...BaseRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.kernelData,
            fields.startPrivateDataTreeSnapshot,
            fields.startNullifierTreeSnapshot,
            fields.startContractTreeSnapshot,
            fields.startPublicDataTreeRoot,
            fields.startHistoricBlocksTreeSnapshot,
            fields.lowNullifierLeafPreimages,
            fields.lowNullifierMembershipWitness,
            fields.newCommitmentsSubtreeSiblingPath,
            fields.newNullifiersSubtreeSiblingPath,
            fields.newContractsSubtreeSiblingPath,
            fields.newPublicDataUpdateRequestsSiblingPaths,
            fields.newPublicDataReadsSiblingPaths,
            fields.historicBlocksTreeRootMembershipWitnesses,
            fields.constants,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...BaseRollupInputs.getFields(this));
    }
}
