// @aztec/circuits.js@0.7.10/dest/structs/rollup/root_rollup.js
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { AggregationObject } from '../aggregation_object.js';
import { GlobalVariables } from '../global_variables.js';
import { AppendOnlyTreeSnapshot } from './append_only_tree_snapshot.js';
/**
 * Represents inputs of the root rollup circuit.
 */
export class RootRollupInputs {
    constructor(
    /**
     * The previous rollup data.
     * Note: Root rollup circuit is the latest circuit the chain of circuits and the previous rollup data is the data
     * from 2 merge or base rollup circuits.
     */
    previousRollupData, 
    /**
     * New L1 to L2 messages.
     */
    newL1ToL2Messages, 
    /**
     * Sibling path of the new L1 to L2 message tree root.
     */
    newL1ToL2MessagesTreeRootSiblingPath, 
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    startL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the historic block roots tree at the start of the rollup.
     */
    startHistoricBlocksTreeSnapshot, 
    /**
     * Sibling path of the new historic block roots tree root.
     */
    newHistoricBlocksTreeSiblingPath) {
        this.previousRollupData = previousRollupData;
        this.newL1ToL2Messages = newL1ToL2Messages;
        this.newL1ToL2MessagesTreeRootSiblingPath = newL1ToL2MessagesTreeRootSiblingPath;
        this.startL1ToL2MessagesTreeSnapshot = startL1ToL2MessagesTreeSnapshot;
        this.startHistoricBlocksTreeSnapshot = startHistoricBlocksTreeSnapshot;
        this.newHistoricBlocksTreeSiblingPath = newHistoricBlocksTreeSiblingPath;
    }
    toBuffer() {
        return serializeToBuffer(...RootRollupInputs.getFields(this));
    }
    static from(fields) {
        return new RootRollupInputs(...RootRollupInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.previousRollupData,
            fields.newL1ToL2Messages,
            fields.newL1ToL2MessagesTreeRootSiblingPath,
            fields.startL1ToL2MessagesTreeSnapshot,
            fields.startHistoricBlocksTreeSnapshot,
            fields.newHistoricBlocksTreeSiblingPath,
        ];
    }
}
/**
 * Represents public inputs of the root rollup circuit.
 *
 * NOTE: in practice, we'll hash all of this up into a single public input, for cheap on-chain verification.
 */
export class RootRollupPublicInputs {
    constructor(
    /**
     * Native aggregation state at the end of the rollup.
     */
    endAggregationObject, 
    /**
     * Global variables of the L2 block.
     */
    globalVariables, 
    /**
     * Snapshot of the private data tree at the start of the rollup.
     */
    startPrivateDataTreeSnapshot, 
    /**
     * Snapshot of the private data tree at the end of the rollup.
     */
    endPrivateDataTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the start of the rollup.
     */
    startNullifierTreeSnapshot, 
    /**
     * Snapshot of the nullifier tree at the end of the rollup.
     */
    endNullifierTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the start of the rollup.
     */
    startContractTreeSnapshot, 
    /**
     * Snapshot of the contract tree at the end of the rollup.
     */
    endContractTreeSnapshot, 
    /**
     * Root of the public data tree at the start of the rollup.
     */
    startPublicDataTreeRoot, 
    /**
     * Root of the public data tree at the end of the rollup.
     */
    endPublicDataTreeRoot, 
    /**
     * Snapshot of the historic private data tree roots tree at the start of the rollup.
     */
    startTreeOfHistoricPrivateDataTreeRootsSnapshot, 
    /**
     * Snapshot of the historic private data tree roots tree at the end of the rollup.
     */
    endTreeOfHistoricPrivateDataTreeRootsSnapshot, 
    /**
     * Snapshot of the historic contract tree roots tree at the start of the rollup.
     */
    startTreeOfHistoricContractTreeRootsSnapshot, 
    /**
     * Snapshot of the historic contract tree roots tree at the end of the rollup.
     */
    endTreeOfHistoricContractTreeRootsSnapshot, 
    /**
     * Snapshot of the L1 to L2 message tree at the start of the rollup.
     */
    startL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the L1 to L2 message tree at the end of the rollup.
     */
    endL1ToL2MessagesTreeSnapshot, 
    /**
     * Snapshot of the historic L1 to L2 message tree roots tree at the start of the rollup.
     */
    startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot, 
    /**
     * Snapshot of the historic L1 to L2 message tree roots tree at the end of the rollup.
     */
    endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot, 
    /**
     * Snapshot of the historic blocks tree roots tree at the start of the rollup.
     */
    startHistoricBlocksTreeSnapshot, 
    /**
     * Snapshot of the historic blocks tree roots tree at the end of the rollup.
     */
    endHistoricBlocksTreeSnapshot, 
    /**
     * Hash of the calldata.
     */
    calldataHash, 
    /**
     * Hash of the L1 to L2 messages.
     */
    l1ToL2MessagesHash) {
        this.endAggregationObject = endAggregationObject;
        this.globalVariables = globalVariables;
        this.startPrivateDataTreeSnapshot = startPrivateDataTreeSnapshot;
        this.endPrivateDataTreeSnapshot = endPrivateDataTreeSnapshot;
        this.startNullifierTreeSnapshot = startNullifierTreeSnapshot;
        this.endNullifierTreeSnapshot = endNullifierTreeSnapshot;
        this.startContractTreeSnapshot = startContractTreeSnapshot;
        this.endContractTreeSnapshot = endContractTreeSnapshot;
        this.startPublicDataTreeRoot = startPublicDataTreeRoot;
        this.endPublicDataTreeRoot = endPublicDataTreeRoot;
        this.startTreeOfHistoricPrivateDataTreeRootsSnapshot = startTreeOfHistoricPrivateDataTreeRootsSnapshot;
        this.endTreeOfHistoricPrivateDataTreeRootsSnapshot = endTreeOfHistoricPrivateDataTreeRootsSnapshot;
        this.startTreeOfHistoricContractTreeRootsSnapshot = startTreeOfHistoricContractTreeRootsSnapshot;
        this.endTreeOfHistoricContractTreeRootsSnapshot = endTreeOfHistoricContractTreeRootsSnapshot;
        this.startL1ToL2MessagesTreeSnapshot = startL1ToL2MessagesTreeSnapshot;
        this.endL1ToL2MessagesTreeSnapshot = endL1ToL2MessagesTreeSnapshot;
        this.startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot = startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot;
        this.endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot = endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot;
        this.startHistoricBlocksTreeSnapshot = startHistoricBlocksTreeSnapshot;
        this.endHistoricBlocksTreeSnapshot = endHistoricBlocksTreeSnapshot;
        this.calldataHash = calldataHash;
        this.l1ToL2MessagesHash = l1ToL2MessagesHash;
    }
    static getFields(fields) {
        return [
            fields.endAggregationObject,
            fields.globalVariables,
            fields.startPrivateDataTreeSnapshot,
            fields.endPrivateDataTreeSnapshot,
            fields.startNullifierTreeSnapshot,
            fields.endNullifierTreeSnapshot,
            fields.startContractTreeSnapshot,
            fields.endContractTreeSnapshot,
            fields.startPublicDataTreeRoot,
            fields.endPublicDataTreeRoot,
            fields.startTreeOfHistoricPrivateDataTreeRootsSnapshot,
            fields.endTreeOfHistoricPrivateDataTreeRootsSnapshot,
            fields.startTreeOfHistoricContractTreeRootsSnapshot,
            fields.endTreeOfHistoricContractTreeRootsSnapshot,
            fields.startL1ToL2MessagesTreeSnapshot,
            fields.endL1ToL2MessagesTreeSnapshot,
            fields.startTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot,
            fields.endTreeOfHistoricL1ToL2MessagesTreeRootsSnapshot,
            fields.startHistoricBlocksTreeSnapshot,
            fields.endHistoricBlocksTreeSnapshot,
            fields.calldataHash,
            fields.l1ToL2MessagesHash,
        ];
    }
    toBuffer() {
        return serializeToBuffer(...RootRollupPublicInputs.getFields(this));
    }
    static from(fields) {
        return new RootRollupPublicInputs(...RootRollupPublicInputs.getFields(fields));
    }
    /**
     * Returns the sha256 hash of the calldata.
     * @returns The sha256 hash of the calldata.
     */
    sha256CalldataHash() {
        const high = this.calldataHash[0].toBuffer();
        const low = this.calldataHash[1].toBuffer();
        const hash = Buffer.alloc(32);
        for (let i = 0; i < 16; i++) {
            hash[i] = high[i + 16];
            hash[i + 16] = low[i + 16];
        }
        return hash;
    }
    /**
     * Deserializes a buffer into a `RootRollupPublicInputs` object.
     * @param buffer - The buffer to deserialize.
     * @returns The deserialized `RootRollupPublicInputs` object.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new RootRollupPublicInputs(reader.readObject(AggregationObject), reader.readObject(GlobalVariables), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readFr(), reader.readFr(), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), [reader.readFr(), reader.readFr()], [reader.readFr(), reader.readFr()]);
    }
}
