// @aztec/circuits.js@0.7.10/dest/structs/public_circuit_public_inputs.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { isArrayEmpty } from '@aztec/foundation/collection';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { MAX_NEW_COMMITMENTS_PER_CALL, MAX_NEW_L2_TO_L1_MSGS_PER_CALL, MAX_NEW_NULLIFIERS_PER_CALL, MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, MAX_PUBLIC_DATA_READS_PER_CALL, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL, RETURN_VALUES_LENGTH, } from '../cbind/constants.gen.js';
import { makeTuple } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { CallContext } from './call_context.js';
import { HistoricBlockData } from './index.js';
/**
 * Contract storage read operation on a specific contract.
 *
 * Note: Similar to `PublicDataRead` but it's from the POV of contract storage so we are not working with public data
 * tree leaf index but storage slot index.
 */
export class ContractStorageRead {
    constructor(
    /**
     * Storage slot we are reading from.
     */
    storageSlot, 
    /**
     * Value read from the storage slot.
     */
    currentValue, 
    /**
     * Optional side effect counter tracking position of this event in tx execution.
     */
    sideEffectCounter) {
        this.storageSlot = storageSlot;
        this.currentValue = currentValue;
        this.sideEffectCounter = sideEffectCounter;
    }
    static from(args) {
        return new ContractStorageRead(args.storageSlot, args.currentValue, args.sideEffectCounter);
    }
    toBuffer() {
        return serializeToBuffer(this.storageSlot, this.currentValue);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ContractStorageRead(reader.readFr(), reader.readFr());
    }
    static empty() {
        return new ContractStorageRead(Fr.ZERO, Fr.ZERO);
    }
    isEmpty() {
        return this.storageSlot.isZero() && this.currentValue.isZero();
    }
    toFriendlyJSON() {
        return `Slot=${this.storageSlot.toFriendlyJSON()}: ${this.currentValue.toFriendlyJSON()}`;
    }
}
/**
 * Contract storage update request for a slot on a specific contract.
 *
 * Note: Similar to `PublicDataUpdateRequest` but it's from the POV of contract storage so we are not working with
 * public data tree leaf index but storage slot index.
 */
export class ContractStorageUpdateRequest {
    constructor(
    /**
     * Storage slot we are updating.
     */
    storageSlot, 
    /**
     * Old value of the storage slot.
     */
    oldValue, 
    /**
     * New value of the storage slot.
     */
    newValue, 
    /**
     * Optional side effect counter tracking position of this event in tx execution.
     */
    sideEffectCounter) {
        this.storageSlot = storageSlot;
        this.oldValue = oldValue;
        this.newValue = newValue;
        this.sideEffectCounter = sideEffectCounter;
    }
    toBuffer() {
        return serializeToBuffer(this.storageSlot, this.oldValue, this.newValue);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ContractStorageUpdateRequest(reader.readFr(), reader.readFr(), reader.readFr());
    }
    /**
     * Create PublicCallRequest from a fields dictionary.
     * @param fields - The dictionary.
     * @returns A PublicCallRequest object.
     */
    static from(fields) {
        return new ContractStorageUpdateRequest(...ContractStorageUpdateRequest.getFields(fields));
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [fields.storageSlot, fields.oldValue, fields.newValue, fields.sideEffectCounter];
    }
    static empty() {
        return new ContractStorageUpdateRequest(Fr.ZERO, Fr.ZERO, Fr.ZERO);
    }
    isEmpty() {
        return this.storageSlot.isZero() && this.oldValue.isZero() && this.newValue.isZero();
    }
    toFriendlyJSON() {
        return `Slot=${this.storageSlot.toFriendlyJSON()}: ${this.oldValue.toFriendlyJSON()} => ${this.newValue.toFriendlyJSON()}`;
    }
}
/**
 * Public inputs to a public circuit.
 */
export class PublicCircuitPublicInputs {
    constructor(
    /**
     * Current call context.
     */
    callContext, 
    /**
     * Pedersen hash of the arguments of the call.
     */
    argsHash, 
    /**
     * Return values of the call.
     */
    returnValues, 
    /**
     * Contract storage update requests executed during the call.
     */
    contractStorageUpdateRequests, 
    /**
     * Contract storage reads executed during the call.
     */
    contractStorageReads, 
    /**
     * Public call stack of the current kernel iteration.
     */
    publicCallStack, 
    /**
     * New commitments created within a public execution call
     */
    newCommitments, 
    /**
     * New nullifiers created within a public execution call
     */
    newNullifiers, 
    /**
     * New L2 to L1 messages generated during the call.
     */
    newL2ToL1Msgs, 
    /**
     * Hash of the unencrypted logs emitted in this function call.
     * Note: Represented as an array of 2 fields in order to fit in all of the 256 bits of sha256 hash.
     */
    unencryptedLogsHash, 
    /**
     * Length of the unencrypted log preimages emitted in this function call.
     */
    unencryptedLogPreimagesLength, 
    /**
     * Root of the commitment trees when the call started.
     */
    historicBlockData, 
    /**
     * Address of the prover.
     */
    proverAddress) {
        this.callContext = callContext;
        this.argsHash = argsHash;
        this.returnValues = returnValues;
        this.contractStorageUpdateRequests = contractStorageUpdateRequests;
        this.contractStorageReads = contractStorageReads;
        this.publicCallStack = publicCallStack;
        this.newCommitments = newCommitments;
        this.newNullifiers = newNullifiers;
        this.newL2ToL1Msgs = newL2ToL1Msgs;
        this.unencryptedLogsHash = unencryptedLogsHash;
        this.unencryptedLogPreimagesLength = unencryptedLogPreimagesLength;
        this.historicBlockData = historicBlockData;
        this.proverAddress = proverAddress;
    }
    /**
     * Create PublicCircuitPublicInputs from a fields dictionary.
     * @param fields - The dictionary.
     * @returns A PublicCircuitPublicInputs object.
     */
    static from(fields) {
        return new PublicCircuitPublicInputs(...PublicCircuitPublicInputs.getFields(fields));
    }
    /**
     * Returns an empty instance.
     * @returns An empty instance.
     */
    static empty() {
        return new PublicCircuitPublicInputs(CallContext.empty(), Fr.ZERO, makeTuple(RETURN_VALUES_LENGTH, Fr.zero), makeTuple(MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL, ContractStorageUpdateRequest.empty), makeTuple(MAX_PUBLIC_DATA_READS_PER_CALL, ContractStorageRead.empty), makeTuple(MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, Fr.zero), makeTuple(MAX_NEW_COMMITMENTS_PER_CALL, Fr.zero), makeTuple(MAX_NEW_NULLIFIERS_PER_CALL, Fr.zero), makeTuple(MAX_NEW_L2_TO_L1_MSGS_PER_CALL, Fr.zero), makeTuple(2, Fr.zero), Fr.ZERO, HistoricBlockData.empty(), AztecAddress.ZERO);
    }
    isEmpty() {
        const isFrArrayEmpty = (arr) => isArrayEmpty(arr, item => item.isZero());
        return (this.callContext.isEmpty() &&
            this.argsHash.isZero() &&
            isFrArrayEmpty(this.returnValues) &&
            isArrayEmpty(this.contractStorageUpdateRequests, item => item.isEmpty()) &&
            isArrayEmpty(this.contractStorageReads, item => item.isEmpty()) &&
            isFrArrayEmpty(this.publicCallStack) &&
            isFrArrayEmpty(this.newCommitments) &&
            isFrArrayEmpty(this.newNullifiers) &&
            isFrArrayEmpty(this.newL2ToL1Msgs) &&
            isFrArrayEmpty(this.unencryptedLogsHash) &&
            this.unencryptedLogPreimagesLength.isZero() &&
            this.historicBlockData.isEmpty() &&
            this.proverAddress.isZero());
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            fields.callContext,
            fields.argsHash,
            fields.returnValues,
            fields.contractStorageUpdateRequests,
            fields.contractStorageReads,
            fields.publicCallStack,
            fields.newCommitments,
            fields.newNullifiers,
            fields.newL2ToL1Msgs,
            fields.unencryptedLogsHash,
            fields.unencryptedLogPreimagesLength,
            fields.historicBlockData,
            fields.proverAddress,
        ];
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PublicCircuitPublicInputs.getFields(this));
    }
}
