// @aztec/foundation@0.7.10/dest/log/logger.js
import debug from 'debug';
import isNode from 'detect-node';
import { isatty } from 'tty';
// Matches a subset of Winston log levels
const LogLevels = ['silent', 'error', 'warn', 'info', 'verbose', 'debug'];
const DefaultLogLevel = 'info';
const envLogLevel = process.env.LOG_LEVEL?.toLowerCase();
const currentLevel = LogLevels.includes(envLogLevel) ? envLogLevel : DefaultLogLevel;
/**
 * Creates a new DebugLogger for the current module, defaulting to the LOG_LEVEL env var.
 * If DEBUG="[module]" env is set, will enable debug logging if the module matches.
 * Uses npm debug for debug level and console.error for other levels.
 * @param name - Name of the module.
 * @returns A debug logger.
 */
export function createDebugLogger(name) {
    const debugLogger = debug(name);
    if (currentLevel === 'debug')
        debugLogger.enabled = true;
    const logger = {
        silent: () => { },
        error: (...args) => logWithDebug(debugLogger, 'error', args),
        warn: (...args) => logWithDebug(debugLogger, 'warn', args),
        info: (...args) => logWithDebug(debugLogger, 'info', args),
        verbose: (...args) => logWithDebug(debugLogger, 'verbose', args),
        debug: (...args) => logWithDebug(debugLogger, 'debug', args),
    };
    return Object.assign((...args) => logWithDebug(debugLogger, 'debug', args), logger);
}
const logHandlers = [];
/**
 * Registers a callback for all logs, whether they are emitted in the current log level or not.
 * @param handler - Callback to be called on every log.
 */
export function onLog(handler) {
    logHandlers.push(handler);
}
/**
 * Logs args to npm debug if enabled or log level is debug, console.error otherwise.
 * @param debug - Instance of npm debug.
 * @param level - Intended log level.
 * @param args - Args to log.
 */
function logWithDebug(debug, level, args) {
    for (const handler of logHandlers) {
        handler(level, debug.namespace, args);
    }
    if (debug.enabled) {
        debug(args[0], ...args.slice(1));
    }
    else if (LogLevels.indexOf(level) <= LogLevels.indexOf(currentLevel) && process.env.NODE_ENV !== 'test') {
        printLog([getPrefix(debug, level), ...args]);
    }
}
/**
 * Returns a log prefix that emulates that of npm debug. Uses colors if in node and in a tty.
 * @param debugLogger - Instance of npm debug logger.
 * @param level - Intended log level (printed out if strictly above current log level).
 * @returns Log prefix.
 */
function getPrefix(debugLogger, level) {
    const levelLabel = currentLevel !== level ? ` ${level.toUpperCase()}` : '';
    const prefix = `${debugLogger.namespace.replace(/^aztec:/, '')}${levelLabel}`;
    if (!isNode || !isatty(process.stderr.fd))
        return prefix;
    const colorIndex = debug.selectColor(debugLogger.namespace);
    const colorCode = '\u001B[3' + (colorIndex < 8 ? colorIndex : '8;5;' + colorIndex);
    return `  ${colorCode};1m${prefix}\u001B[0m`;
}
/**
 * Outputs to console error.
 * @param args - Args to log.
 */
function printLog(args) {
    // eslint-disable-next-line no-console
    console.error(...args);
}
