// @aztec/types@0.7.10/dest/logs/note_spending_info/note_spending_info.js
import { AztecAddress } from '@aztec/circuits.js';
import { serializeToBuffer } from '@aztec/circuits.js/utils';
import { Fr, GrumpkinScalar } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { decryptBuffer, encryptBuffer } from './encrypt_buffer.js';
import { NotePreimage } from './note_preimage.js';
/**
 * A class which wraps the data required to compute a nullifier/to spend a note. Along with that this class contains
 * the necessary functionality to encrypt and decrypt the data.
 */
export class NoteSpendingInfo {
    constructor(
    /**
     * Preimage which can be used along with private key to compute nullifier.
     */
    notePreimage, 
    /**
     * Address of the contract this tx is interacting with.
     */
    contractAddress, 
    /**
     * Storage slot of the contract this tx is interacting with.
     */
    storageSlot) {
        this.notePreimage = notePreimage;
        this.contractAddress = contractAddress;
        this.storageSlot = storageSlot;
    }
    /**
     * Deserializes the NoteSpendingInfo object from a Buffer.
     * @param buffer - Buffer or BufferReader object to deserialize.
     * @returns An instance of NoteSpendingInfo.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new NoteSpendingInfo(reader.readObject(NotePreimage), reader.readObject(AztecAddress), reader.readFr());
    }
    /**
     * Serializes the NoteSpendingInfo object into a Buffer.
     * @returns Buffer representation of the NoteSpendingInfo object.
     */
    toBuffer() {
        return serializeToBuffer([this.notePreimage, this.contractAddress, this.storageSlot]);
    }
    /**
     * Encrypt the NoteSpendingInfo object using the owner's public key and the ephemeral private key.
     * @param ownerPubKey - Public key of the owner of the NoteSpendingInfo object.
     * @param curve - The curve instance to use.
     * @returns The encrypted NoteSpendingInfo object.
     */
    toEncryptedBuffer(ownerPubKey, curve) {
        const ephPrivKey = GrumpkinScalar.random();
        return encryptBuffer(this.toBuffer(), ownerPubKey, ephPrivKey, curve);
    }
    /**
     * Decrypts the NoteSpendingInfo object using the owner's private key.
     * @param data - Encrypted NoteSpendingInfo object.
     * @param ownerPrivKey - Private key of the owner of the NoteSpendingInfo object.
     * @param curve - The curve instance to use.
     * @returns Instance of NoteSpendingInfo if the decryption was successful, undefined otherwise.
     */
    static fromEncryptedBuffer(data, ownerPrivKey, curve) {
        const buf = decryptBuffer(data, ownerPrivKey, curve);
        if (!buf) {
            return;
        }
        return NoteSpendingInfo.fromBuffer(buf);
    }
    /**
     * Create a random NoteSpendingInfo object (useful for testing purposes).
     * @returns A random NoteSpendingInfo object.
     */
    static random() {
        return new NoteSpendingInfo(NotePreimage.random(), AztecAddress.random(), Fr.random());
    }
}
