// @aztec/noir-contracts@0.7.10/dest/artifacts/index.js
import CardGameJson from './card_game_contract.json' assert { type: 'json' };
export const CardGameContractAbi = CardGameJson;
import ChildJson from './child_contract.json' assert { type: 'json' };
export const ChildContractAbi = ChildJson;
import DocsExampleJson from './docs_example_contract.json' assert { type: 'json' };
export const DocsExampleContractAbi = DocsExampleJson;
import EasyPrivateTokenJson from './easy_private_token_contract.json' assert { type: 'json' };
export const EasyPrivateTokenContractAbi = EasyPrivateTokenJson;
import EcdsaAccountJson from './ecdsa_account_contract.json' assert { type: 'json' };
export const EcdsaAccountContractAbi = EcdsaAccountJson;
import EscrowJson from './escrow_contract.json' assert { type: 'json' };
export const EscrowContractAbi = EscrowJson;
import ImportTestJson from './import_test_contract.json' assert { type: 'json' };
export const ImportTestContractAbi = ImportTestJson;
import LendingJson from './lending_contract.json' assert { type: 'json' };
export const LendingContractAbi = LendingJson;
import MultiTransferJson from './multi_transfer_contract.json' assert { type: 'json' };
export const MultiTransferContractAbi = MultiTransferJson;
import NonNativeTokenJson from './non_native_token_contract.json' assert { type: 'json' };
export const NonNativeTokenContractAbi = NonNativeTokenJson;
import ParentJson from './parent_contract.json' assert { type: 'json' };
export const ParentContractAbi = ParentJson;
import PendingCommitmentsJson from './pending_commitments_contract.json' assert { type: 'json' };
export const PendingCommitmentsContractAbi = PendingCommitmentsJson;
import PokeableTokenJson from './pokeable_token_contract.json' assert { type: 'json' };
export const PokeableTokenContractAbi = PokeableTokenJson;
import PriceFeedJson from './price_feed_contract.json' assert { type: 'json' };
export const PriceFeedContractAbi = PriceFeedJson;
import PrivateTokenAirdropJson from './private_token_airdrop_contract.json' assert { type: 'json' };
export const PrivateTokenAirdropContractAbi = PrivateTokenAirdropJson;
import PrivateTokenJson from './private_token_contract.json' assert { type: 'json' };
export const PrivateTokenContractAbi = PrivateTokenJson;
import PublicTokenJson from './public_token_contract.json' assert { type: 'json' };
export const PublicTokenContractAbi = PublicTokenJson;
import SchnorrAccountJson from './schnorr_account_contract.json' assert { type: 'json' };
export const SchnorrAccountContractAbi = SchnorrAccountJson;
import SchnorrHardcodedAccountJson from './schnorr_hardcoded_account_contract.json' assert { type: 'json' };
export const SchnorrHardcodedAccountContractAbi = SchnorrHardcodedAccountJson;
import SchnorrSingleKeyAccountJson from './schnorr_single_key_account_contract.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractAbi = SchnorrSingleKeyAccountJson;
import StatefulTestJson from './stateful_test_contract.json' assert { type: 'json' };
export const StatefulTestContractAbi = StatefulTestJson;
import TestJson from './test_contract.json' assert { type: 'json' };
export const TestContractAbi = TestJson;
import TokenBridgeJson from './token_bridge_contract.json' assert { type: 'json' };
export const TokenBridgeContractAbi = TokenBridgeJson;
import TokenJson from './token_contract.json' assert { type: 'json' };
export const TokenContractAbi = TokenJson;
import UniswapJson from './uniswap_contract.json' assert { type: 'json' };
export const UniswapContractAbi = UniswapJson;
