// @aztec/aztec.js@0.7.10/dest/contract/contract_base.js
import { FunctionSelector } from '@aztec/foundation/abi';
import { ContractFunctionInteraction } from './contract_function_interaction.js';
/**
 * Abstract implementation of a contract extended by the Contract class and generated contract types.
 */
export class ContractBase {
    constructor(
    /**
     * The deployed contract's complete address.
     */
    completeAddress, 
    /**
     * The Application Binary Interface for the contract.
     */
    abi, 
    /**
     * The wallet.
     */
    wallet) {
        this.completeAddress = completeAddress;
        this.abi = abi;
        this.wallet = wallet;
        /**
         * An object containing contract methods mapped to their respective names.
         */
        this.methods = {};
        abi.functions.forEach((f) => {
            const interactionFunction = (...args) => {
                return new ContractFunctionInteraction(this.wallet, this.completeAddress.address, f, args);
            };
            this.methods[f.name] = Object.assign(interactionFunction, {
                /**
                 * A getter for users to fetch the function selector.
                 * @returns Selector of the function.
                 */
                get selector() {
                    return FunctionSelector.fromNameAndParameters(f.name, f.parameters);
                },
            });
        });
    }
    /**
     * Address of the contract.
     */
    get address() {
        return this.completeAddress.address;
    }
    /**
     * Creates a new instance of the contract wrapper attached to a different wallet.
     * @param wallet - Wallet to use for sending txs.
     * @returns A new contract instance.
     */
    withWallet(wallet) {
        return new ContractBase(this.completeAddress, this.abi, wallet);
    }
    /**
     * Attach the current contract instance to a portal contract and optionally add its dependencies.
     * The function will return a promise that resolves when all contracts have been added to the AztecRPCClient.
     * This is useful when you need to interact with a deployed contract that has multiple nested contracts.
     *
     * @param portalContract - The Ethereum address of the portal contract.
     * @param dependencies - An optional array of additional DeployedContract instances to be attached.
     * @returns A promise that resolves when all contracts are successfully added to the AztecRPCClient.
     */
    attach(portalContract, dependencies = []) {
        const deployedContract = {
            abi: this.abi,
            completeAddress: this.completeAddress,
            portalContract,
        };
        return this.wallet.addContracts([deployedContract, ...dependencies]);
    }
}
