// @aztec/circuits.js@0.7.10/dest/structs/kernel/private_kernel.js
import { serializeToBuffer } from '../../utils/serialize.js';
/**
 * Private call data.
 * @see circuits/cpp/src/aztec3/circuits/abis/call_stack_item.hpp
 */
export class PrivateCallData {
    constructor(
    /**
     * The call stack item currently being processed.
     */
    callStackItem, 
    /**
     * Other private call stack items to be processed.
     */
    privateCallStackPreimages, 
    /**
     * The proof of the execution of this private call.
     */
    proof, 
    /**
     * The verification key for the function being invoked.
     */
    vk, 
    /**
     * The membership witness for the function leaf corresponding to the function being invoked.
     */
    functionLeafMembershipWitness, 
    /**
     * The membership witness for the contract leaf corresponding to the contract on which the function is being
     * invoked.
     */
    contractLeafMembershipWitness, 
    /**
     * The membership witnesses for read requests created by the function being invoked.
     */
    readRequestMembershipWitnesses, 
    /**
     * The address of the portal contract corresponding to the contract on which the function is being invoked.
     */
    portalContractAddress, 
    /**
     * The hash of the ACIR of the function being invoked.
     */
    acirHash) {
        this.callStackItem = callStackItem;
        this.privateCallStackPreimages = privateCallStackPreimages;
        this.proof = proof;
        this.vk = vk;
        this.functionLeafMembershipWitness = functionLeafMembershipWitness;
        this.contractLeafMembershipWitness = contractLeafMembershipWitness;
        this.readRequestMembershipWitnesses = readRequestMembershipWitnesses;
        this.portalContractAddress = portalContractAddress;
        this.acirHash = acirHash;
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            // NOTE: Must have same order as CPP.
            fields.callStackItem,
            fields.privateCallStackPreimages,
            fields.proof,
            fields.vk,
            fields.functionLeafMembershipWitness,
            fields.contractLeafMembershipWitness,
            fields.readRequestMembershipWitnesses,
            fields.portalContractAddress,
            fields.acirHash,
        ];
    }
    static from(fields) {
        return new PrivateCallData(...PrivateCallData.getFields(fields));
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PrivateCallData.getFields(this));
    }
}
/**
 * Input to the private kernel circuit - initial call.
 */
export class PrivateKernelInputsInit {
    constructor(
    /**
     * The transaction request which led to the creation of these inputs.
     */
    txRequest, 
    /**
     * Private calldata corresponding to this iteration of the kernel.
     */
    privateCall) {
        this.txRequest = txRequest;
        this.privateCall = privateCall;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.txRequest, this.privateCall);
    }
}
/**
 * Input to the private kernel circuit - Inner call.
 */
export class PrivateKernelInputsInner {
    constructor(
    /**
     * The previous kernel data (dummy if this is the first kernel).
     */
    previousKernel, 
    /**
     * Private calldata corresponding to this iteration of the kernel.
     */
    privateCall) {
        this.previousKernel = previousKernel;
        this.privateCall = privateCall;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.privateCall);
    }
}
/**
 * Input to the private kernel circuit - Final ordering call.
 */
export class PrivateKernelInputsOrdering {
    constructor(
    /**
     * The previous kernel data
     */
    previousKernel, 
    /**
     * Contains hints for the transient read requests to localize corresponding commitments.
     */
    readCommitmentHints, 
    /**
     * Contains hints for the transient nullifiers to localize corresponding commitments.
     */
    nullifierCommitmentHints) {
        this.previousKernel = previousKernel;
        this.readCommitmentHints = readCommitmentHints;
        this.nullifierCommitmentHints = nullifierCommitmentHints;
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.readCommitmentHints);
    }
}
