// @aztec/circuits.js@0.7.10/dest/structs/kernel/public_inputs.js
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../../utils/serialize.js';
import { CombinedAccumulatedData } from './combined_accumulated_data.js';
import { CombinedConstantData } from './combined_constant_data.js';
/**
 * Public inputs of the public and private kernel circuits.
 * @see circuits/cpp/src/aztec3/circuits/abis/kernel_circuit_public_inputs.hpp
 */
export class KernelCircuitPublicInputs {
    constructor(
    /**
     * Data accumulated from both public and private circuits.
     */
    end, 
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Indicates whether the input is for a private or public kernel.
     */
    isPrivate) {
        this.end = end;
        this.constants = constants;
        this.isPrivate = isPrivate;
    }
    toBuffer() {
        return serializeToBuffer(this.end, this.constants, this.isPrivate);
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of KernelCircuitPublicInputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new KernelCircuitPublicInputs(reader.readObject(CombinedAccumulatedData), reader.readObject(CombinedConstantData), reader.readBoolean());
    }
    static empty() {
        return new KernelCircuitPublicInputs(CombinedAccumulatedData.empty(), CombinedConstantData.empty(), true);
    }
}
/**
 * Public inputs of the public kernel circuit.
 */
export class PublicKernelPublicInputs extends KernelCircuitPublicInputs {
    constructor(end, constants) {
        super(end, constants, false);
    }
    static empty() {
        return new PublicKernelPublicInputs(CombinedAccumulatedData.empty(), CombinedConstantData.empty());
    }
}
/**
 * Public inputs of the private kernel circuit.
 */
export class PrivateKernelPublicInputs extends KernelCircuitPublicInputs {
    constructor(end, constants) {
        super(end, constants, true);
    }
    static empty() {
        return new PrivateKernelPublicInputs(CombinedAccumulatedData.empty(), CombinedConstantData.empty());
    }
}
