// @aztec/circuits.js@0.7.10/dest/structs/kernel/public_kernel.js
import { MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, MAX_PUBLIC_DATA_READS_PER_TX, MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX, } from '../../cbind/constants.gen.js';
import { assertMemberLength } from '../../utils/jsUtils.js';
import { serializeToBuffer } from '../../utils/serialize.js';
/**
 * Inputs to the public kernel circuit.
 */
export class PublicKernelInputs {
    constructor(
    /**
     * Kernels are recursive and this is the data from the previous kernel.
     */
    previousKernel, 
    /**
     * Public calldata assembled from the execution result and proof.
     */
    publicCall) {
        this.previousKernel = previousKernel;
        this.publicCall = publicCall;
    }
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.publicCall);
    }
}
/**
 // eslint-disable-next-line tsdoc/syntax
 * TODO: POSSIBLY OBSOLETE --\> DELETE OR DOCUMENT.
 */
export class WitnessedPublicCallData {
    constructor(
    /**
     * TODO.
     */
    publicCall, 
    /**
     * TODO.
     */
    updateRequestsHashPaths, 
    /**
     * TODO.
     */
    readsHashPaths, 
    /**
     * TODO.
     */
    publicDataTreeRoot) {
        this.publicCall = publicCall;
        this.updateRequestsHashPaths = updateRequestsHashPaths;
        this.readsHashPaths = readsHashPaths;
        this.publicDataTreeRoot = publicDataTreeRoot;
        assertMemberLength(this, 'updateRequestsHashPaths', MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX);
        assertMemberLength(this, 'readsHashPaths', MAX_PUBLIC_DATA_READS_PER_TX);
    }
    toBuffer() {
        return serializeToBuffer(this.publicCall, this.updateRequestsHashPaths, this.readsHashPaths, this.publicDataTreeRoot);
    }
}
/**
 * Public calldata assembled from the kernel execution result and proof.
 */
export class PublicCallData {
    constructor(
    /**
     * Call stack item being processed by the current iteration of the kernel.
     */
    callStackItem, 
    /**
     * Children call stack items.
     */
    publicCallStackPreimages, 
    /**
     * Proof of the call stack item execution.
     */
    proof, 
    /**
     * Address of the corresponding portal contract.
     */
    portalContractAddress, 
    /**
     * Hash of the L2 contract bytecode.
     */
    bytecodeHash) {
        this.callStackItem = callStackItem;
        this.publicCallStackPreimages = publicCallStackPreimages;
        this.proof = proof;
        this.portalContractAddress = portalContractAddress;
        this.bytecodeHash = bytecodeHash;
        assertMemberLength(this, 'publicCallStackPreimages', MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL);
    }
    toBuffer() {
        return serializeToBuffer(this.callStackItem, this.publicCallStackPreimages, this.proof, this.portalContractAddress, this.bytecodeHash);
    }
}
