// @aztec/aztec.js@0.7.10/dest/wallet/base_wallet.js
/**
 * A base class for Wallet implementations
 */
export class BaseWallet {
    constructor(rpc) {
        this.rpc = rpc;
    }
    registerAccount(privKey, partialAddress) {
        return this.rpc.registerAccount(privKey, partialAddress);
    }
    registerRecipient(account) {
        return this.rpc.registerRecipient(account);
    }
    getRegisteredAccounts() {
        return this.rpc.getRegisteredAccounts();
    }
    getRegisteredAccount(address) {
        return this.rpc.getRegisteredAccount(address);
    }
    getRecipients() {
        return this.rpc.getRecipients();
    }
    getRecipient(address) {
        return this.rpc.getRecipient(address);
    }
    addContracts(contracts) {
        return this.rpc.addContracts(contracts);
    }
    getContracts() {
        return this.rpc.getContracts();
    }
    simulateTx(txRequest, simulatePublic) {
        return this.rpc.simulateTx(txRequest, simulatePublic);
    }
    sendTx(tx) {
        return this.rpc.sendTx(tx);
    }
    getTxReceipt(txHash) {
        return this.rpc.getTxReceipt(txHash);
    }
    getPrivateStorageAt(owner, contract, storageSlot) {
        return this.rpc.getPrivateStorageAt(owner, contract, storageSlot);
    }
    getPublicStorageAt(contract, storageSlot) {
        return this.rpc.getPublicStorageAt(contract, storageSlot);
    }
    getNoteNonces(contract, storageSlot, preimage, txHash) {
        return this.rpc.getNoteNonces(contract, storageSlot, preimage, txHash);
    }
    viewTx(functionName, args, to, from) {
        return this.rpc.viewTx(functionName, args, to, from);
    }
    getExtendedContractData(contractAddress) {
        return this.rpc.getExtendedContractData(contractAddress);
    }
    getContractData(contractAddress) {
        return this.rpc.getContractData(contractAddress);
    }
    getUnencryptedLogs(from, limit) {
        return this.rpc.getUnencryptedLogs(from, limit);
    }
    getBlockNumber() {
        return this.rpc.getBlockNumber();
    }
    getNodeInfo() {
        return this.rpc.getNodeInfo();
    }
    isGlobalStateSynchronised() {
        return this.rpc.isGlobalStateSynchronised();
    }
    isAccountStateSynchronised(account) {
        return this.rpc.isAccountStateSynchronised(account);
    }
    getSyncStatus() {
        return this.rpc.getSyncStatus();
    }
    addAuthWitness(authWitness) {
        return this.rpc.addAuthWitness(authWitness);
    }
}
