// @aztec/types@0.7.10/dest/l2_block_downloader/l2_block_downloader.js
import { MemoryFifo, Semaphore, SerialQueue } from '@aztec/foundation/fifo';
import { createDebugLogger } from '@aztec/foundation/log';
import { InterruptableSleep } from '@aztec/foundation/sleep';
import { INITIAL_L2_BLOCK_NUM } from '../index.js';
const log = createDebugLogger('aztec:l2_block_downloader');
/**
 * Downloads L2 blocks from a L2BlockSource.
 * The blocks are stored in a queue and can be retrieved using the getBlocks method.
 * The queue size is limited by the maxQueueSize parameter.
 * The downloader will pause when the queue is full or when the L2BlockSource is out of blocks.
 */
export class L2BlockDownloader {
    constructor(l2BlockSource, maxQueueSize, pollIntervalMS = 10000) {
        this.l2BlockSource = l2BlockSource;
        this.pollIntervalMS = pollIntervalMS;
        this.running = false;
        this.from = 0;
        this.interruptableSleep = new InterruptableSleep();
        this.jobQueue = new SerialQueue();
        this.blockQueue = new MemoryFifo();
        this.semaphore = new Semaphore(maxQueueSize);
    }
    /**
     * Starts the downloader.
     * @param from - The block number to start downloading from. Defaults to INITIAL_L2_BLOCK_NUM.
     */
    start(from = INITIAL_L2_BLOCK_NUM) {
        if (this.running) {
            this.interruptableSleep.interrupt();
            return;
        }
        this.from = from;
        this.running = true;
        const fn = async () => {
            while (this.running) {
                try {
                    await this.jobQueue.put(() => this.collectBlocks());
                    await this.interruptableSleep.sleep(this.pollIntervalMS);
                }
                catch (err) {
                    log.error(err);
                    await this.interruptableSleep.sleep(this.pollIntervalMS);
                }
            }
        };
        this.jobQueue.start();
        this.runningPromise = fn();
    }
    /**
     * Repeatedly queries the block source and adds the received blocks to the block queue.
     * Stops when no further blocks are received.
     * @returns The total number of blocks added to the block queue.
     */
    async collectBlocks() {
        let totalBlocks = 0;
        while (true) {
            const blocks = await this.l2BlockSource.getL2Blocks(this.from, 10);
            if (!blocks.length) {
                return totalBlocks;
            }
            await this.semaphore.acquire();
            this.blockQueue.put(blocks);
            this.from += blocks.length;
            totalBlocks += blocks.length;
        }
    }
    /**
     * Stops the downloader.
     */
    async stop() {
        this.running = false;
        this.interruptableSleep.interrupt();
        await this.jobQueue.cancel();
        this.blockQueue.cancel();
        await this.runningPromise;
    }
    /**
     * Gets the next batch of blocks from the queue.
     * @param timeout - optional timeout value to prevent permanent blocking
     * @returns The next batch of blocks from the queue.
     */
    async getL2Blocks(timeout) {
        try {
            const blocks = await this.blockQueue.get(timeout);
            if (!blocks) {
                return [];
            }
            this.semaphore.release();
            return blocks;
        }
        catch (err) {
            // nothing to do
            return [];
        }
    }
    /**
     * Forces an immediate request for blocks.
     * @returns A promise that fulfills once the poll is complete
     */
    pollImmediate() {
        return this.jobQueue.put(() => this.collectBlocks());
    }
}
