// @aztec/noir-contracts@0.7.10/src/artifacts/index.ts
// Auto generated module\n
import { ContractAbi } from '@aztec/foundation/abi';
import CardGameJson from './card_game_contract.json' assert { type: 'json' };
export const CardGameContractAbi = CardGameJson as ContractAbi;
import ChildJson from './child_contract.json' assert { type: 'json' };
export const ChildContractAbi = ChildJson as ContractAbi;
import DocsExampleJson from './docs_example_contract.json' assert { type: 'json' };
export const DocsExampleContractAbi = DocsExampleJson as ContractAbi;
import EasyPrivateTokenJson from './easy_private_token_contract.json' assert { type: 'json' };
export const EasyPrivateTokenContractAbi = EasyPrivateTokenJson as ContractAbi;
import EcdsaAccountJson from './ecdsa_account_contract.json' assert { type: 'json' };
export const EcdsaAccountContractAbi = EcdsaAccountJson as ContractAbi;
import EscrowJson from './escrow_contract.json' assert { type: 'json' };
export const EscrowContractAbi = EscrowJson as ContractAbi;
import ImportTestJson from './import_test_contract.json' assert { type: 'json' };
export const ImportTestContractAbi = ImportTestJson as ContractAbi;
import LendingJson from './lending_contract.json' assert { type: 'json' };
export const LendingContractAbi = LendingJson as ContractAbi;
import MultiTransferJson from './multi_transfer_contract.json' assert { type: 'json' };
export const MultiTransferContractAbi = MultiTransferJson as ContractAbi;
import NonNativeTokenJson from './non_native_token_contract.json' assert { type: 'json' };
export const NonNativeTokenContractAbi = NonNativeTokenJson as ContractAbi;
import ParentJson from './parent_contract.json' assert { type: 'json' };
export const ParentContractAbi = ParentJson as ContractAbi;
import PendingCommitmentsJson from './pending_commitments_contract.json' assert { type: 'json' };
export const PendingCommitmentsContractAbi = PendingCommitmentsJson as ContractAbi;
import PokeableTokenJson from './pokeable_token_contract.json' assert { type: 'json' };
export const PokeableTokenContractAbi = PokeableTokenJson as ContractAbi;
import PriceFeedJson from './price_feed_contract.json' assert { type: 'json' };
export const PriceFeedContractAbi = PriceFeedJson as ContractAbi;
import PrivateTokenAirdropJson from './private_token_airdrop_contract.json' assert { type: 'json' };
export const PrivateTokenAirdropContractAbi = PrivateTokenAirdropJson as ContractAbi;
import PrivateTokenJson from './private_token_contract.json' assert { type: 'json' };
export const PrivateTokenContractAbi = PrivateTokenJson as ContractAbi;
import PublicTokenJson from './public_token_contract.json' assert { type: 'json' };
export const PublicTokenContractAbi = PublicTokenJson as ContractAbi;
import SchnorrAccountJson from './schnorr_account_contract.json' assert { type: 'json' };
export const SchnorrAccountContractAbi = SchnorrAccountJson as ContractAbi;
import SchnorrHardcodedAccountJson from './schnorr_hardcoded_account_contract.json' assert { type: 'json' };
export const SchnorrHardcodedAccountContractAbi = SchnorrHardcodedAccountJson as ContractAbi;
import SchnorrSingleKeyAccountJson from './schnorr_single_key_account_contract.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractAbi = SchnorrSingleKeyAccountJson as ContractAbi;
import StatefulTestJson from './stateful_test_contract.json' assert { type: 'json' };
export const StatefulTestContractAbi = StatefulTestJson as ContractAbi;
import TestJson from './test_contract.json' assert { type: 'json' };
export const TestContractAbi = TestJson as ContractAbi;
import TokenBridgeJson from './token_bridge_contract.json' assert { type: 'json' };
export const TokenBridgeContractAbi = TokenBridgeJson as ContractAbi;
import TokenJson from './token_contract.json' assert { type: 'json' };
export const TokenContractAbi = TokenJson as ContractAbi;
import UniswapJson from './uniswap_contract.json' assert { type: 'json' };
export const UniswapContractAbi = UniswapJson as ContractAbi;
