// @aztec/aztec.js@0.7.10/dest/contract_deployer/contract_deployer.js
import { Point } from '@aztec/circuits.js';
import { DeployMethod } from './deploy_method.js';
/**
 * A class for deploying contract.
 * @remarks Keeping this around even though we have Aztec.nr contract types because it can be useful for non-TS users.
 */
export class ContractDeployer {
    constructor(abi, arc, publicKey) {
        this.abi = abi;
        this.arc = arc;
        this.publicKey = publicKey;
    }
    /**
     * Deploy a contract using the provided ABI and constructor arguments.
     * This function creates a new DeployMethod instance that can be used to send deployment transactions
     * and query deployment status. The method accepts any number of constructor arguments, which will
     * be passed to the contract's constructor during deployment.
     *
     * @param args - The constructor arguments for the contract being deployed.
     * @returns A DeployMethod instance configured with the ABI, AztecRPCClient, and constructor arguments.
     */
    deploy(...args) {
        return new DeployMethod(this.publicKey ?? Point.ZERO, this.arc, this.abi, args);
    }
}
